File: src/film/cartoon/cartoon.cpp
#include "cartoon.h"

#include "../../utils/io/io.h"
#include "../../utils/random/random.h"
#include <cstdio>
#include <cstdlib>
#include <cstring>


char *Map(enum Cartoon::type type) {
  char *result = nullptr;
  if (type == Cartoon::DRAWING) {
    result = "DRAWING";
  } else if (type == Cartoon::PUPPET) {
    result = "PUPPET";
  } else if (type == Cartoon::PLASTICINE) {
    result = "PLASTICINE";
  }

  return result;
}

enum Cartoon::type Map(char *type) {
  if (strcmp(type, "DRAWING") == 0) {
    return Cartoon::DRAWING;
  } else if (strcmp(type, "PUPPET") == 0) {
    return Cartoon::PUPPET;
  } else if (strcmp(type, "PLASTICINE") == 0) {
    return Cartoon::PLASTICINE;
  } else {
    printf("Incorrect type of Cartoon film given: %s", type);
    exit(1);
  }
}

enum Cartoon::type Map(int typeNumber) {
  if (typeNumber == 1) {
    return Cartoon::DRAWING;
  } else if (typeNumber == 2) {
    return Cartoon::PUPPET;
  } else if (typeNumber == 3) {
    return Cartoon::PLASTICINE;
  } else {
    printf("Incorrect type number of Cartoon film generated by random: %d", typeNumber);
    exit(2);
  }
}


Film *Cartoon::InStochastic() {
  return new Cartoon(
      RandomString(RandomInt(1, 51)),
      RandomInt(1970, 3022),
      Map(RandomInt(1, 4)));
}

Film *Cartoon::In(const char *title, int year, FILE *input) {
  return new Cartoon(title, year, Map(ReadString(input)));
}

void Cartoon::Out(FILE *output) {
  fprintf(output, "Cartoon{title=%s,year=%d,type=%s}", title, year, Map(type));
}

Cartoon::Cartoon(const char *title, int year, enum Cartoon::type type) : Film(title, year), type(type) {}

File: src/film/cartoon/cartoon.py
import random

from src.film.film import Film

from enum import Enum

from src.utils import io


class CartoonGenre(Enum):
    DRAWING = "DRAWING"
    PUPPET = "PUPPET"
    PLASTICINE = "PLASTICINE"


class Cartoon(Film):
    def __init__(self):
        super().__init__()
        self.genre = None

    def read_from_file(self, file):
        super(Cartoon, self).read_from_file(file)
        self.genre = io.read_enum(file, CartoonGenre, "Cartoon genre")

    def fill_stochastic(self):
        super(Cartoon, self).fill_stochastic()
        self.genre = random.choice(list(CartoonGenre))[0]

    def write_to_file(self, file):
        file.write('Cartoon{' + F'title={self.title},year={self.year},genre={self.genre}' + '}')
